PL2/PL2-core/src/main/java/nl/tudelft/pl2016gr2/core/algorithms/FilterBubbles.java

```diff
@@ -57,189 +46,34 @@ public FilterBubbles(Collection<GraphNode> orderedNodes) {
-      visited.add(next);
-      poppedNodes.add(next);
-      List<GraphNode> bubbleLinks = new ArrayList<>();
-      for (GraphNode outlink : calcNodeOutlinks(originalOutEdges.get(next.getId()),
-          leaves, bubble)) {
-        GraphNode node = outlink;
-        if (FilterHelpers.isShared(node, leaves)) {
-          FilterHelpers.addToVisit(outlink, toVisit, visited);
-        } else {
-          bubbleLinks.add(outlink);
-        }
-      }
-      createBubble(treeNode, next, bubbleLinks, leaves, toVisit, visited, newBubbles, poppedNodes);
-    }
-  }
-
-  private void addNestedNodes(List<GraphNode> nestedNodes, Bubble bubble) {
-    for (GraphNode node : nestedNodes) {
-      BubbleChildrenVisitor visitor = new BubbleChildrenVisitor(node);
-      bubble.accept(visitor);
-    }
-  }
-
-  private void createBubble(IPhylogeneticTreeNode treeNode, GraphNode inlink,
-      List<GraphNode> bubbleLinks, List<Integer> leaves, Queue<GraphNode> toVisit,
-      Set<GraphNode> visited, List<Bubble> newBubbles, Set<GraphNode> poppedNodes) {
-    if (!bubbleLinks.isEmpty()) {
-      Bubble newBubble = new PhyloBubble(mutationId, treeNode, this);
-      if (inlink != null) {
-        newBubble.addInEdge(inlink);
-      }
-      addNestedNodes(bubbleLinks, newBubble);
-      Queue<GraphNode> toVisitNodes = makeBubble(newBubble, bubbleLinks, leaves);
-      while (!toVisitNodes.isEmpty()) {
-        FilterHelpers.addToVisit(toVisitNodes.poll(), toVisit, visited);
-      }
-
-      newBubbles.add(newBubble);
-      poppedNodes.add(newBubble);
-      mutationId--;
-    }
-  }
-
-  private Queue<GraphNode> makeBubble(Bubble bubble, List<GraphNode> start, List<Integer> leaves) {
-    Queue<GraphNode> endPoints = new LinkedList<>();
-    Set<GraphNode> visited = new HashSet<>();
-    Queue<GraphNode> toVisit = new LinkedList<>();
-
-    toVisit.addAll(start);
-
-    while (!toVisit.isEmpty()) {
-      GraphNode next = toVisit.poll();
-      visited.add(next);
-
-      if (FilterHelpers.isShared(next, leaves)) {
-        endPoints.add(next);
-        bubble.addOutEdge(next);
-      } else {
-        bubble.accept(new BubbleChildrenVisitor(next));
-
-        for (GraphNode outlink : next.getOutEdges()) {
-          FilterHelpers.addToVisit(outlink, toVisit, visited);
-        }
-      }
-    }
-
-    return endPoints;
-  }
-
-  private List<GraphNode> calcNodeOutlinks(Collection<GraphNode> outEdges,
-      ArrayList<Integer> leaves, Bubble bubble) {
-    List<GraphNode> curNodeOutlinks = new ArrayList<>();
-
-    for (GraphNode outlink : outEdges) {
-      if (bubble != null && !bubble.hasChild(outlink)) {
-        continue;
-      }
-      ArrayList<Integer> genomes = new ArrayList<>(outlink.getGenomes());
-
-      for (Integer leaf : leaves) {
-        if (genomes.contains(leaf)) {
-          curNodeOutlinks.add(outlink);
-          break;
-        }
-      }
-    }
-
-    return curNodeOutlinks;
+  protected Collection<GraphNode> getRootNodes() {
```

Comment:
I don't have access to the SequenceGraph, so I can't store it. 

I don't see why storing the root nodes is a problem, because they won't change (because the root nodes of the graph before it's been filtered by this class are stored). And I don't need to check whether nodes are a root node, I just need the root nodes for my algorithm. 

I could instead save the orderedNodes list that is given is as input and then calculate the rootNodes on the fly, but that would mean that I would need to save this list. Also, this list could change, so basing the rootNodes on this would not be very secure.